### PR TITLE
Extract files by node version order && support 'force' arg to overwri…

### PR DESCRIPTION
**Pull request by gorgiaxx**
According to this document. The blocks should be sorted by `inode.version`. 
https://sourceware.org/jffs2/jffs2-slides-transformed.pdf

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
+*.py[cod]
```
Comment (qkaiser) [inline .gitignore:3]: This is out of scope of this pull request. Either remove it or move it into a dedicated commit.

File: src/scripts/jefferson [hunk 2]
```diff
@@ -54,13 +54,6 @@ JFFS2_NODETYPE_XREF = JFFS2_FEATURE_INCOMPAT | JFFS2_NODE_ACCURATE | 9
 def mtd_crc(data):
     return (binascii.crc32(data, -1) ^ -1) & 0xFFFFFFFF
 
-def is_safe_path(basedir, path, follow_symlinks=True):
-    if follow_symlinks:
-        matchpath = os.path.realpath(path)
-    else:
-        matchpath = os.path.abspath(path)
-    return basedir == os.path.commonpath((basedir, matchpath))
-
```
Comment (qkaiser) [inline src/scripts/jefferson:63]: Why would you remove this ? It was introduced in https://github.com/sviehb/jefferson/commit/41f654eac078cfec240428be457c054613a80851 to protect againt path traversal through malicious JFFS2.

File: src/scripts/jefferson [hunk 3]
```diff
@@ -89,6 +82,57 @@ class Jffs2_unknown_node(cstruct.CStruct):
...
+            jint16_t magic;
+            jint16_t nodetype;      /* = JFFS2_NODETYPE_XREF */
+            jint32_t totlen;
+            jint32_t hdr_crc;
+            jint32_t ino;           /* inode number */
+            jint32_t xid;           /* XATTR identifier number */
+            jint32_t xseqno;        /* xref sequencial number */
+            jint32_t node_crc;
+        }
+    """
```
Comment (qkaiser) [inline src/scripts/jefferson:135]: Why re-introduce those structures ? Quoting the commit messages, we removed them because:

> These nodes are never used to reconstruct the filesystem so we can just skip them to make jefferson faster.

File: src/scripts/jefferson [hunk 4]
```diff
@@ -265,6 +311,23 @@ def set_endianness(endianness):
...
+        Jffs2_raw_xattr.__def__,
+        __name__=Jffs2_raw_xattr.__name__,
+        __byte_order__=endianness,
+    )
+
+    Jffs2_raw_xref = Jffs2_raw_xref.parse(
+        Jffs2_raw_xref.__def__,
+        __name__=Jffs2_raw_xref.__name__,
+        __byte_order__=endianness,
+    )
```
Comment (qkaiser) [inline src/scripts/jefferson:330]: Not needed because those node types are not used to reconstruct the filesystem on disk.

File: src/scripts/jefferson [hunk 5]
```diff
@@ -275,6 +338,9 @@ def scan_fs(content, endianness, verbose=False):
     fs = {}
     fs[JFFS2_NODETYPE_INODE] = {}
     fs[JFFS2_NODETYPE_DIRENT] = {}
+    fs[JFFS2_NODETYPE_XATTR] = {}
+    fs[JFFS2_NODETYPE_XREF] = {}
+    fs[JFFS2_NODETYPE_SUMMARY] = []
```
Comment (qkaiser) [inline src/scripts/jefferson:343]: Not needed because those node types are not used to reconstruct the filesystem on disk.

File: src/scripts/jefferson [hunk 6]
```diff
@@ -318,21 +384,42 @@ def scan_fs(content, endianness, verbose=False):
...
+                    xref = Jffs2_raw_xref()
+                    xref.unpack(content_mv[0 + offset :])
+
+                    if xref.ino in fs[JFFS2_NODETYPE_XREF]:
+                        if xref.version > fs[JFFS2_NODETYPE_XREF][xref.ino].version:
+                            fs[JFFS2_NODETYPE_XREF][xref.ino] = xref
+                    else:
+                        fs[JFFS2_NODETYPE_XREF][xref.ino] = xref
+                    if verbose:
+                        print("0x%08X:" % (offset), xref)
```
Comment (qkaiser) [inline src/scripts/jefferson:419]: Not needed because those node types are not used to reconstruct the filesystem on disk. We parse structures that are never used afterwards.

File: src/scripts/jefferson [hunk 7]
```diff
@@ -318,21 +384,42 @@ def scan_fs(content, endianness, verbose=False):
...
+                    xattr = Jffs2_raw_xattr()
+                    xattr.unpack(content_mv[0 + offset :])
+
+                    if xattr.ino in fs[JFFS2_NODETYPE_XATTR]:
+                        if xattr.version > fs[JFFS2_NODETYPE_XATTR][xattr.ino].version:
+                            fs[JFFS2_NODETYPE_XATTR][xattr.ino] = xattr
+                    else:
+                        fs[JFFS2_NODETYPE_XATTR][xattr.ino] = xattr
+                    if verbose:
+                        print("0x%08X:" % (offset), xattr)
```
Comment (qkaiser) [inline src/scripts/jefferson:408]: Not needed because those node types are not used to reconstruct the filesystem on disk. We parse structures that are never used afterwards.

File: src/scripts/jefferson [hunk 8]
```diff
@@ -318,21 +384,42 @@ def scan_fs(content, endianness, verbose=False):
...
-                    pass
-                elif unknown_node.nodetype == JFFS2_NODETYPE_PADDING:
-                    pass
                 elif unknown_node.nodetype == JFFS2_NODETYPE_SUMMARY:
-                    pass
+                    summary = Jffs2_raw_summary()
+                    summary.unpack(content_mv[0 + offset :])
+                    fs[JFFS2_NODETYPE_SUMMARY].append(summary)
+                    if verbose:
+                        print("0x%08X:" % (offset), summary)
```
Comment (qkaiser) [inline src/scripts/jefferson:397]: Not needed because those node types are not used to reconstruct the filesystem on disk. We parse structures that are never used afterwards.

File: src/scripts/jefferson [hunk 9]
```diff
@@ -389,25 +483,26 @@ def dump_fs(fs, target):
 
         target_path = os.path.realpath(os.path.join(os.getcwd(), target, path))
 
-        if not is_safe_path(target, target_path):
-            print(f"Path traversal attempt to {target_path}, discarding.")
-            continue
```
Comment (qkaiser) [inline src/scripts/jefferson:394]: Any reason behind this removal ?

File: src/scripts/jefferson [hunk 10]
```diff
@@ -485,7 +585,7 @@ def main():
         if not os.path.exists(dest_path):
             os.mkdir(dest_path)
 
-        dump_fs(fs, dest_path)
+        dump_fs(fs, dest_path, verbose=args.verbose, force=args.force)
```
Comment (qkaiser) [inline src/scripts/jefferson:588]: Please move the addition of `verbose` mode into its own commit, same thing for `force` mode. We're trying to use atomic git commits.

**Review by qkaiser — CHANGES_REQUESTED** [review]

**Comment by qkaiser** [comment]
Thanks for submitting your first PR to jefferson, I just did a first review pass :)

File: src/scripts/jefferson (hunk 2 above)
Comment (gorgiaxx) [reply src/scripts/jefferson:63]: I patched this file based on the old commit, so it doesn't include your new code. I'll close this PR

Comment (qkaiser) [reply src/scripts/jefferson:63]: Do you plan on submitting a PR based on the current tree ? Otherwise I can create a PR based on your code fix for version handling. Your call :)

Comment (gorgiaxx) [reply src/scripts/jefferson:63]: That would be better if you create a PR and fix it with some of my code. Because now you are the maintner of jefferson. Your safe patch is incompatible with my "Firmware Analyzer". It reports path traversal errors when I use normal firmware.

**Comment by qkaiser** [comment]
@gorgiaxx I fixed the bug you reported and released version 0.4 so users can track which version to use. It was a regression introduced in a recent commit.

I also created #45 and #46 to keep track of verbosity and force overwrite option that you wanted to introduce with this MR.

Thanks again for reporting this !
